### ISSUE TITLE
Cubic spline interpolation for large data sets

### ISSUE THREAD

**Issue by Simsso** [Simsso/Online-Tools#1]
The cubic spline interpolation tool (http://tools.timodenk.com/cubic-spline-interpolation) produces _wrong_ output for the attached data (see screenshot below). _Thanks to Samir Ouchene for reporting this bug._

The boundary conditions, currently set to **natural spline**, could be causing that issue. Possible solution: Setting the boundary conditions to **not-a-knot spline** as described in the following equation.

![image](https://cloud.githubusercontent.com/assets/6556307/23329821/46d28f9c-fb40-11e6-84f4-f840fdc5a7bb.png)

```JavaScript
// Not-a-knot spline (needs to be adapted - currently second derivative, should be third)
m[row][0+0] = 6*p[1].x;
m[row][0+1] = 2;
m[row][0+4] = -6*p[1].x;
m[row++][0+5] = -2;
m[row][solutionIndex-8+0] = 6*p[p.length - 1].x;
m[row][solutionIndex-8+1] = 2;
m[row][solutionIndex-8+4] = -6*p[p.length - 1].x;
m[row++][solutionIndex-8+5] = -2;
```

![image](https://cloud.githubusercontent.com/assets/6556307/23329777/5e648846-fb3f-11e6-8d08-7deff01cdf56.png)

```
-2.23       -6.00000E-02
-0.1610     1.560000E-01
1.84        3.690000E-01
3.88        5.710000E-01
5.89        7.550000E-01
7.89        8.600000E-01
8.95        8.870000E-01
9.91        8.690000E-01
10.9        8.680000E-01
12          8.940000E-01
12.9        9.380000E-01
14          9.290000E-01
14.9        9.080000E-01
16          9.120000E-01
17          6.550000E-01
18          5.880000E-01
19          5.870000E-01
20          5.970000E-01
22          6.030000E-01
24          6.470000E-01
26          6.830000E-01
28.1        7.450000E-01
30          8.240000E-01
35          1.05
40          1.14
45          1.2
50          1.12
55          1.17
60          1.08
65          9.400000E-01
70          8.570000E-01
74.9        6.660000E-01
79.9        4.720000E-01
84.8        3.560000E-01
89.9        1.420000E-01
```

**Comment by Simsso** [comment]
As Arne Bergseth reported, the behavior is comparable with the following data:
```
-6458 -270
-6404 -250
-5891 -200
-4913 -150
-3554 -100
-1889 -50
0 0
4096 100
6138 150
7340 180
```
![image](https://cloud.githubusercontent.com/assets/6556307/24695925/024123a4-19e8-11e7-97b9-f1b6236a1f29.png)

**Comment by Simsso** [comment]
Three boundary condition types were implemented in the commit [2765a9f2db160b00ac2f712b8f81cc055f10ff0b](https://github.com/Simsso/Online-Tools/commit/2765a9f2db160b00ac2f712b8f81cc055f10ff0b) and the problem still exists. For the first set of data the problem does not occur when all points with x>16 are removed from the input data set. 


**Therefore the JavaScript floating point number precision is likely to be causing this problem.** Usage of a library like [math.js](http://mathjs.org/) should be considered.

![image](https://cloud.githubusercontent.com/assets/6556307/25997813/3a680088-371e-11e7-9aa7-8adc9797939c.png)
